src/csharp/Grpc.Examples.Tests/MockMathClientTest.cs

```diff
@@ -0,0 +1,132 @@
+
+// Copyright 2015, Google Inc.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//     * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//     * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core;
+using Grpc.Core.Utils;
+using Grpc.Core.Testing;
+using NUnit.Framework;
+
+namespace Math.Tests
+{
+    /// <summary>
+    /// Example of mocking a gRPC service client for unit testing purposes.
+    /// </summary>
+    public class MockMathClientTest
+    {
+        FakeMathClient mockClient = new FakeMathClient();
+
+        [Test]
+        public async void MockAsyncUnaryCall()
+        {
+            var expectedResponse = new DivReply
+            {
+                Quotient = 3,
+                Remainder = 1
+            };
+            var mockCall = new MockAsyncUnaryCall<DivReply>();
+            mockCall.SetResponse(expectedResponse);
+
+            mockClient.DivAsyncImpl = (request, options) =>
+            {
+                return mockCall.Call;
+            };
+
+            var response = await mockClient.DivAsync(new DivArgs { Dividend = 10, Divisor = 3 }, new CallOptions());
+            Assert.AreEqual(expectedResponse, response);
+        }
+
+        /// <summary>
+        /// Demonstrates how to create a fake implementation of IMathClient.
+        /// </summary>
+        class FakeMathClient : Math.IMathClient
+        {
+            public Func<DivArgs, CallOptions, DivReply> DivImpl;
```

Comment:
Personally I think it would be simpler to have a sort of adapter that would take an IMath, and proxy the call (still via MockAsyncUnaryCall). I don't think we'd want to differentiate between sync and async in the fake.
But I think I can work with what you've given for MockAsyncUnaryCall to achieve that :)